src/corecel/math/RegulaFalsiRootFinder.hh

```diff
@@ -0,0 +1,110 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020-2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+/*! \file corecel/math/RegulaFalsiRootFinder.hh
+ *
+ * Perform a Regula Falsi Iteration given a root function \em func and
+ * toolerance \em tol .
+ *
+ * Using a \em left and \em right bound a Regula Falsi approximates the \em
+ * root as: \f[ root = (left * func(right) - right * func(left)) / (func(right)
+ * - func(left)) \f]
+ *
+ * Then value of \em func at the root is calculated compared to values of
+ * \em func at the bounds. The bound which has the same sign \em func(root)
+ */
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cmath>
+#include <type_traits>
+
+#include "celeritas_config.h"
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/math/Algorithms.hh"
+
+#include "detail/MathImpl.hh"
+
+namespace celeritas
+{
+template<class F>
+class RegulaFalsi
+{
+  public:
+    // Contructpr of Regula Falsi
+    inline CELER_FUNCTION RegulaFalsi(F&& func, real_type tol);
+
+    // Solve between
+    real_type operator()(real_type left, real_type right);
+
+  private:
+    F&& func_;
+    real_type tol_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+// TEMPLATE DEDUCTION
+//---------------------------------------------------------------------------//
+
+template<class F, class... Args>
+RegulaFalsi(F&&, Args...) -> RegulaFalsi<F>;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from function.
+ */
+template<class F>
+CELER_FUNCTION RegulaFalsi<F>::RegulaFalsi(F&& func, real_type tol)
+    : func_{func}, tol_{tol}
+{
+    CELER_EXPECT(tol > 0);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Perform Regula Falsi in defined bounds \em left and \em right with
+ * parameters defined in \em params.
+ */
+template<class F>
+CELER_FUNCTION real_type RegulaFalsi<F>::operator()(real_type left,
+                                                    real_type right)
+{
+    // Initialize Iteration parameters
+    real_type f_left = func_(left);
+    real_type f_right = func_(right);
+    real_type f_root = 1;
```

Comment:
Since `f(x) can be unbounded`, then perhaps either the loop should be `do {} while ();` or initialized to `numeric_limits<real_type>::infinity()` (`corecel/math/NumericLimits.hh` for the device-compatible version).